[FLINK-20740][network] Introduce a separated buffer pool and a separated thread pool for sort-merge blocking shuffle

wsry opened the pull request:
## What is the purpose of the change

Currently, sort-merge blocking shuffle implementation uses some unmanaged direct memory and the netty thread for shuffle data read which can lead to direct memory OOM and stress the netty thread with file IO. This patch introduces a separated buffer pool and a separated thread pool for data read. In the following patch, the problem will be totally solved by the IO scheduling mechanism leveraging the introduced buffer pool and thread pool.

## Brief change log

  - Introduce a separated buffer pool and a separated thread pool for sort-merge blocking shuffle.


## Verifying this change

This change added tests BatchReadBufferPoolTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

wsry commented:
@flinkbot run azure

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                            && counter.get() + numBuffersPerRequest > maxBuffersPerRequester)) {
+                checkState(!destroyed, "Buffer pool is already destroyed.");
+
+                buffers.wait(requestTimeout);
```
seems `buffers` is not notified when `numBuffersAllocated` is updated.

wsry replied:
Do you mean when the all buffers are recycled and the allocated buffers reaches 0. If that is the case, the timeout will be triggered in the next round. Currently, we only notify when there is enough buffers or the buffer pool is destroyed.

gaoyunhaii replied:
I'm afraid we might still need the notification since it is also waiting for the condition `(counter != null
&& counter.get() + numBuffersPerRequest > maxBuffersPerRequester)`, if there is no notification for this condition, after there are enough buffers, the thread might still blocked on this condition.

wsry replied:
The recycle thread will always notify the lock

zhuzhurk replied:
close this comment because it is outdated with the removal of `maxBuffersPerRequester`

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                buffers.wait(requestTimeout);
+                counter = numBuffersAllocated.get(owner);
+
+                if (counter == null && System.currentTimeMillis() - startTime >= requestTimeout) {
```
should exception also be thrown if `counter != null && counter.get() + numBuffersPerRequest <= maxBuffersPerRequester`?

wsry replied:
It is not guaranteed the requester can get the upper bound number of buffers.

zhuzhurk reviewed (commented):
Thanks for opening this PR @wsry 
I have a few comments. Please take a look.

wsry commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -513,6 +513,30 @@
 
+    /**
+     * Memory used by blocking shuffle for shuffle data read (currently only used by sort-merge
+     * shuffle). The minimum valid value can be configured is 32M.
```
The reason for this is that too small read memory is bad for performance and 32M is not a big memory overhead. Removing this restriction is also acceptable for me. I will remove and add some descriptions like: decreasing this value may impact the performance.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadIOExecutor.java:
```diff
@@ -0,0 +1,127 @@
+public class BatchShuffleReadIOExecutor implements Executor {
+
+    /** Minimum valid number of executor threads. */
+    public static final int MIN_NUM_THREADS = 4;
```
A minimum number of 4 threads is good for IO performance because IO usually need some waiting operation which does not consume cpu cycles. If we only have 1 thread, resources may be wasted. And after the change above, we do not need 32M memory any more.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadIOExecutor.java:
```diff
@@ -0,0 +1,127 @@
+ * used by sort-merge blocking shuffle.
+ */
+@Internal
+public class BatchShuffleReadIOExecutor implements Executor {
```
BatchShuffleReadIOExecutor will be Initialized lazily when first used.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+        this(totalBytes, bufferSize, 5 * 60 * 1000);
+    }
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize, long requestTimeout) {
```
You are right, it only for test currently, because the default 5min timeout is a little long for test. BatchShuffleReadBufferPoolTest#testRequestBuffersTimeout and BatchShuffleReadBufferPoolTest#testNoBufferTimeoutIfHoldingBuffers use it.

zhuzhurk replied:
Let's annotate it as `@VisibleForTesting` and make it package private.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+        this.bufferSize = bufferSize;
+        this.requestTimeout = requestTimeout;
+
+        this.numTotalBuffers = (int) Math.min(totalBytes / bufferSize, Integer.MAX_VALUE);
```
numTotalBuffers must be int, because the size of java collections is Integer, we can not avoid checking it.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                bufferSize);
+
+        checkState(
+                totalBytes >= bufferSize,
```
I add this check to initialize instead of constructor to keep the behavior not changed. Because if one have set the memory segment size to larger than 32M, then after he/she upgrades to the new version, exception will be thrown in the constructor even sort-merge shuffle is not enabled.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                        "Illegal configuration, config value for %s must be no smaller than %s,"
+                                + " please increase %s to at least %s bytes.",
+                        TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key(),
+                        TaskManagerOptions.MEMORY_SEGMENT_SIZE.key(),
```
The last %s should be %d.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                }
+            } catch (OutOfMemoryError outOfMemoryError) {
+                int allocated = buffers.size();
+                buffers.clear();
```
I am sure ```buffer.clear()```is enough for direct memory. GC will do the work.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+    /**
+     * Recycles the target buffer to this buffer pool. This method should never throw any exception.
+     */
+    public void recycle(MemorySegment segment, Object owner) {
```
Could we proxy this method to `recycle(Collections.singletonList(segment), owner)` ?

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+        this.maxBuffersPerRequester =
+                Math.max(
+                        4 * numBuffersPerRequest,
+                        numTotalBuffers / BatchShuffleReadIOExecutor.MAX_NUM_THREADS);
```
As discussed offline, we will remove the upper bound and use 4 * numBuffersPerRequest.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                checkState(!destroyed, "Buffer pool is already destroyed.");
+
+                buffers.wait(requestTimeout);
+                counter = numBuffersAllocated.get(owner);
```
After waiting, do we still need to check whether it is destroyed here ?

wsry replied:
I think not needed. The check is at the start of the while loop. If the pool has been destroyed, the buffers will be cleared and the while condition will be true and the check at the start  of the while loop will be reached.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                Math.min(numTotalBuffers, Math.max(1, NUM_BYTES_PER_REQUEST / bufferSize));
+        this.maxBuffersPerRequester =
+                Math.max(
+                        4 * numBuffersPerRequest,
```
It is only an empirical value (32M) to let each allocator to use more than 8M memory if there are more available buffers.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+        this.numBuffersPerRequest =
+                Math.min(numTotalBuffers, Math.max(1, NUM_BYTES_PER_REQUEST / bufferSize));
+        this.maxBuffersPerRequester =
+                Math.max(
```
As discussed offline, we will remove it.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+
+    /** Account for all the buffers requested per requester. */
+    @GuardedBy("buffers")
+    private final Map<Object, Counter> numBuffersAllocated = new HashMap<>();
```
Because a counter object is more convenient for increase and decrease operation.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+        this.bufferSize = bufferSize;
+        this.requestTimeout = requestTimeout;
+
+        this.numTotalBuffers = (int) Math.min(totalBytes / bufferSize, Integer.MAX_VALUE);
```
It was checked it initialize method

wsry replied:
I totally agree we should not enforce that and there is no such enforcement in the latest version.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,353 @@
+                buffers.wait(requestTimeout);
+                counter = numBuffersAllocated.get(owner);
+
+                if (counter == null && System.currentTimeMillis() - startTime >= requestTimeout) {
```
I have another concern here in that based on the condition for waiting, it seems a requester (owner) might have multiple threads work for it concurrently, right ? if so, suppose the current thread keep waiting due to another thread work for it holds too much buffer than `maxBuffersPerRequester` for a long time, and at last round, the other thread recycled all the buffers after timeout, then TimeoutException is thrown here. Would this also be expected ? It seems each requester has hold some buffers in this case and no requester are starved.

wsry replied:
It should be guaranteed only one thread can serve a requester at the same time, I can add this constraint to the java doc of this method.

wsry replied:
buffer recycle will always notify the waiting thread

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleServiceFactory.java:
```diff
@@ -119,13 +121,34 @@ static NettyShuffleEnvironment createNettyShuffleEnvironment(
+        // contention, for example, too long IO or waiting time leading to starvation or timeout
+        int numThreads =
+                batchShuffleReadBufferPool.getNumTotalBuffers()
+                        / Math.max(1, batchShuffleReadBufferPool.getNumBuffersPerRequest());
```
numTotalBuffers can be 0 if the total bytes is less than the buffer size

wsry commented:
@zhuzhurk @gaoyunhaii Thanks for the review and comments. I have updated the PR by pushing a fixup commit. You can take a look.
BTW, I removed the max buffers per requester logic in the fixup commit because the blocking wait occupies a thread which should be used by others. I will implement a similar logic at requester side. (A requester will not request new buffers if too many buffers are already allocated, only when total requested buffers are smaller than the threshold, the data read task will be resubmitted again)

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleServiceFactory.java:
```diff
@@ -119,13 +121,33 @@ static NettyShuffleEnvironment createNettyShuffleEnvironment(
+        // contention, for example, too long IO or waiting time leading to starvation or timeout
+        BatchShuffleReadIOExecutor batchShuffleReadIOExecutor =
+                new BatchShuffleReadIOExecutor(
+                        Math.max(
```
maybe let's do this min and max in the ctor of BatchShuffleReadIOExecutor? Then `BatchShuffleReadIOExecutor.MIN_NUM_THREADS` and `BatchShuffleReadIOExecutor.MAX_NUM_THREADS` can be private.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,275 @@
+        return totalBytes;
+    }
+
+    @VisibleForTesting
```
Seems all these `VisibleForTesting` getters can be package private

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,275 @@
+                checkState(!destroyed, "Buffer pool is already destroyed.");
+
+                buffers.wait(requestTimeout);
+                if (System.currentTimeMillis() - startTime >= requestTimeout) {
```
seems the timeout will never trigger if available `buffers` does not increase to be larger than `numBuffersPerRequest`, because `buffers` is only notified when it can fulfill a request.
I feel that we should not let the timeout rely on buffer returning events, because in the worst case that no buffer can be returned, timeout will never be triggered.

wsry replied:
The wait method has a timeout

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironmentBuilder.java:
```diff
@@ -180,7 +187,8 @@ public NettyShuffleEnvironment build() {
                         sortShuffleMinBuffers,
-                        sortShuffleMinParallelism),
+                        sortShuffleMinParallelism,
+                        batchShuffleReadBuffers),
```
I guess it should be `batchShuffleReadBytes` and it should be placed in ahead of the param `sortShuffleMinBuffers`.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionBuilder.java:
```diff
@@ -50,6 +52,12 @@
     private NetworkBufferPool networkBufferPool = new NetworkBufferPool(2, 1);
 
+    private BatchShuffleReadBufferPool batchShuffleReadBufferPool =
+            new BatchShuffleReadBufferPool(1024, 32 * 1024);
```
why set a `totalBytes` to be smaller than `bufferSize` in this default batchShuffleReadBufferPool? Seems it will lead to exceptions if used.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPoolTest.java:
```diff
@@ -0,0 +1,286 @@
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
```
could we have this test case to test normal case only and have a separate test case `testRequestBuffersAfterPoolDestroyed` to test `requestBuffers` after destroy?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPoolTest.java:
```diff
@@ -0,0 +1,286 @@
+        assertTrue(bufferPool.isDestroyed());
+        assertEquals(0, bufferPool.getAvailableBuffers());
+
+        bufferPool.recycle(buffers, this);
```
maybe add a separate test `testRecycleAfterPoolDestroyed` to test that recycled buffers will be directly discarded if the pool is destroyed?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPoolTest.java:
```diff
@@ -0,0 +1,286 @@
+            assertEquals(bufferPool.getNumBuffersPerRequest(), buffers.size());
+        } finally {
+            bufferPool.recycle(buffers, this);
+            assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
```
this check looks irrelevant to this test case. Can we remove it and add a separate test case `testRecycle` to test the normal simple cycle case, because currently I can only see some complex recycle test cases?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPoolTest.java:
```diff
@@ -0,0 +1,286 @@
+            bufferPool.requestBuffers(new Object());
+        } finally {
+            bufferPool.recycle(buffers, this);
+            assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
```
same as the above comment. Let's have a separate test case for this check.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadIOExecutor.java:
```diff
@@ -0,0 +1,127 @@
+        this.numThreads = numThreads;
+    }
+
+    public int getNumThreads() {
```
it's not needed I guess.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadIOExecutor.java:
```diff
@@ -0,0 +1,127 @@
+     * Initializes this executor pool which creates all executor threads, the executor pool must be
+     * initialized before used.
+     */
+    public void initialize() {
```
I think it can be package private and then we do not need a test for it. `testInitializeAfterDestroy` can be covered by `testExecuteTaskAfterDestroy`.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    /** Initializes this buffer pool which allocates all the buffers. */
+    public void initialize() {
```
Could we make it private and `testInitializeAfterDestroy` can be replaced with a `testRequestBuffersAfterDestroy`?
For those case that need the pool to be initialized, we can use `execute` to trigger the initialization.

wsry commented:
@zhuzhurk Thanks for the review and comments, I have updated the PR.

zhuzhurk reviewed (approved):
Thanks for addressing all the comments. @wsry 
The change looks good to me.

gaoyunhaii commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,264 @@
+                for (int i = 0; i < numTotalBuffers; ++i) {
+                    buffers.add(MemorySegmentFactory.allocateUnpooledOffHeapMemory(bufferSize));
+                }
+            } catch (OutOfMemoryError outOfMemoryError) {
```
I think we might still have a call to `MemorySegment.free()` in case there might be future changes on how memorySegmentFactory allocates off-heap memory on OOM and destroy. We might have a method to cleanup the current `buffer` array to be called in the two places.

xintongsong commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -513,6 +513,28 @@
+                            String.format(
+                                    "Size of memory used by blocking shuffle for shuffle data read "
+                                            + "(currently only used by sort-merge shuffle). Notes: "
+                                            + "1) The memory is cut from %s so must be smaller than"
+                                            + " that, which means you may also need to increase %s "
```
nit:
```suggestion
                                            + "1) The memory is cut from '%s' so must be smaller than"
                                            + " that, which means you may also need to increase '%s' "
```

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,265 @@
+        checkState(
+                totalBytes >= bufferSize,
+                String.format(
+                        "Illegal configuration, config value for %s must be no smaller than %s,"
+                                + " please increase %s to at least %d bytes.",
```
nit:
```suggestion
                        "Illegal configuration, config value for '%s' must be no smaller than '%s',"
                                + " please increase '%s' to at least %d bytes.",
```

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,265 @@
+                            String.format(
+                                    "Can't allocate enough buffers in the given timeout, which means"
+                                            + " there is a fierce contention for read buffers, please"
+                                            + " increase %s.",
```
nit:
```suggestion
                                            + " increase '%s'.",
```

xintongsong reviewed (approved):
I can only speak for the memory configurations, and the PR LGTM from that perspective.

StephanEwen reviewed (commented):
Thanks for this effort, I think this goes in a good direction. Some general thoughts below, and some detail comments inline.

## Lazy initialization

Both the memory pool and the executor use lazy initialization. In my experience, initialization should be done eagerly by default, and only when there is a very specific reason it should by lazy.

I see that there is a good case for the memory pool to make memory allocation lazy, because many applications (streaming only) will never need that memory.

For the executor, I would recommend to initialize it eagerly. The underlying ExecutorService already handles lazy thread spawning and termination, so there isn't any gain by creating it lazy. It just makes the code more complicated an error-prone, and makes config failures surface too late.

## Buffer Request Batching

The code tries to allocate 8 MB of buffers each time, and blocks if less data is available.
I am curious if you have experience how that works in practice. With the default pool size, this means 4 read operations can be in progress at the same time, which sounds really low.

Maybe you have some experimental experience already that shows this is a good approach. If not, I am wondering if we should start without that optimization, and let the pool serve one buffer at a time, see how that works, and use it as a baseline. Then add improvements for more sequential I/O on top of this.

Intuitively, I would also say that the responsibility to offer more sequential I/O is part of the component that accepts read requests and serves the data, not the buffer pool. That component will be able to do more elaborate things, like elevator-algorithm, etc.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,265 @@
+    /** Destroys this buffer pool and after which, no buffer can be allocated any more. */
+    public void destroy() {
+        synchronized (buffers) {
+            destroyed = true;
```
It has been already cleared.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,265 @@
+
+        checkState(
+                totalBytes >= bufferSize,
+                String.format(
```
You are right, I will replace it with if statement though the initialize method is not called frequently.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,265 @@
+                numTotalBuffers,
+                bufferSize);
+
+        checkState(
```
Doing the check in the initialize method is to avoid potential failure for streaming only user.

StephanEwen replied:
If the defaults are fine, then streaming-only users should also never see problems (I wouldn't expect them to change the defaults).

wsry replied:
I agree that it should be rare for the streaming to config the memory segment size too large. I will move the check to constructor.

wsry commented:
@StephanEwen Thanks for the review and comments.

**About Lazy Initialization**
I agree that initializing the ExecutorService eagerly is better if it already handles lazy thread spawning and termination.

**About Buffer Request Batching**
I have already tested the default configuration with tpcds benchmark. Compared with my previous implementation which uses exclusive buffers and exclusive thread per result partition, there is no performance regression. Besides, batch reading is important for IO performance especially on HDD. **Buffer Request Batching** is a simple way to guarantee that the reader can always get enough buffers for sequential read, otherwise, all readers may get a few buffers and read in parallel which can be bad for performance.

wsry commented:
@StephanEwen I have updated the PR.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java:
```diff
@@ -0,0 +1,265 @@
+    @GuardedBy("buffers")
+    private boolean initialized;
+
+    public BatchShuffleReadBufferPool(long totalBytes, int bufferSize, long requestTimeout) {
```
@StephanEwen Sorry for the misunderstanding of this reply, let us just ignore this. I misunderstood the context.

wsry commented:
Hi @StephanEwen, is there any other concerns? If not, we will merge the PR soon.

StephanEwen commented:
Looks good to me.
Can you confirm that when you start a Flink application or cluster with this change, it has no impact at this point?

If yes, then +1 to merge.

wsry commented:
Yes, I can confirm that it has no impact at this point.

wsry commented:
Rebase master and rerun tests.

zhuzhurk commented:
Thanks for the further reviewing! @StephanEwen  @gaoyunhaii 
Merging.
